OpenSim/Actuators/VandenBogert2011Muscle.cpp

```diff
@@ -0,0 +1,1234 @@
+
+/*
+    for (int i =0; i<=1; i=i+1) {
+        yTemp[i]=y[i]+h;
+        del = calcImplicitResidual(yTemp,ydot,u,0);
+        df_dy[0][i]=(del.forceResidual-opForceResidual)/h; //force residual
+        df_dy[1][i]=(del.activResidual-opActivResidual)/h; //activ residual
+
+
+        ydotTemp[i]=y[i]+h;
+        del = calcImplicitResidual(y,ydotTemp,u,0);
+
+        df_dydot[0][i]=(del.forceResidual-opForceResidual)/h; //force residual
+        df_dydot[1][i]=(del.activResidual-opActivResidual)/h; //activ residual
+
+        yTemp=y;
+        ydotTemp=ydot;
+
+    }*/
+
+
+    //del = calcImplicitResidual(y,ydot,u+h,0);
+    //double df_du = (del.forceResidual-opForceResidual)/h;
+
+    opPoint.df_dy=df_dy;
+    opPoint.df_dydot=df_dydot;
+    opPoint.df_du=df_du;
+
+    return opPoint;
+
+}
+
+
+//------------------------------------------------------------------------------
+SimTK::Vec3 VandenBogert2011Muscle::calcFiberStaticEquilibResidual(
+        double projFibLen, double muscleLength, double activ) const
+{
+// Just a quick convience method to get the force residual
+// under static conditions
+
+    VandenBogert2011Muscle::ImplicitResidual Results = calcImplicitResidual(
+            muscleLength, projFibLen, activ, 0.0, 0.0, activ, 1.0);
+
+    SimTK::Vec3 resAndDerivative;
+
+    resAndDerivative[0]=Results.forceResidual;
+    resAndDerivative[1]=Results.df_dy[0][0];
+    resAndDerivative[2]=Results.forceTendon;
+
+    return resAndDerivative;
+
+}
+
+//------------------------------------------------------------------------------
+SimTK::Vec2 VandenBogert2011Muscle::calcFiberStaticEquilbirum(
+        double muscleLength, double activ) const {
+
+    // This code assumes that the muscle lengthing speed is 0.
+
+
+    //TODO: Code is not optimized.  Specifically the number of calls to
+    //calcImplicitResidual can be reduced
+
+    //TODO: Generalize with a Lambda function (will need help with that).
+    //TODO: calcImplicitResidual really only needs to calculate df_ds
+    //       (single element) for this function
+
+
+double tol=1e-8; //TODO : Look at changing to propotional of eps
+double a=0;
+double b=10;  //10 muscle lengths (more will slow convergence by adding steps)
+
+double x=(a+b)/2.0;
+double dx=2*tol;
+
+int neval=0;
+
+SimTK::Vec3 forceResAndDerivative;
+
+    //Perform a Newton Line Search
+while ((abs(dx)>=tol) && (neval<100)) {
+
+    neval++;
+
+    // Set a to be lower value and b to be upper value
+    a=min(a,b);
+    b=max(a,b);
+
+
+    forceResAndDerivative =
+            calcFiberStaticEquilibResidual(x, muscleLength, activ);
+    double fx = forceResAndDerivative[0];
+
+    //After the 1st iteration, use the new guess as a new upper or lower bound
+    if (neval>1) {
+        forceResAndDerivative =
+                calcFiberStaticEquilibResidual(a, muscleLength, activ);
+        double funcA = forceResAndDerivative[0];
+
+        if ((funcA *fx)>0){
+            a = x;}
+        else { b = x;}}
```

Comment:
Is something like this OK:
`        if (~inInterval || largeDeriv) {
            x = (a+b)/2;
            dx = (a-b)/2;}
        else { x = xdx; };
        };`